File: graph/src/languages/test/fixtures/shop.ts
import { EventEmitter } from 'events';
import type { Repository } from './types';
import * as utils from './utils';

export type ItemId = number;
export type Price = { amount: number; currency: string };

export const DEFAULT_TIMEOUT = 5000;
export const MAX_RETRIES = 3;

export enum ItemStatus {
    Pending = 'PENDING',
    Available = 'AVAILABLE',
    Discontinued = 'DISCONTINUED',
}

export interface Serializable<T> {
    serialize(): T;
    deserialize(data: T): void;
}

export interface Cacheable {
    getCacheKey(): string;
    invalidate(): void;
}

export interface Item extends Serializable<Record<string, unknown>> {
    id: ItemId;
    name: string;
    price: Price;
    quantity: number;
    status: ItemStatus;
    tags?: string[];
}

export type ItemFactory<T extends Item = Item> = (data: Partial<T>) => T;

export abstract class BaseEntity {
    protected id: number;
    private _createdAt: Date;

    constructor(id: number) {
        this.id = id;
        this._createdAt = new Date();
    }

    abstract validate(): boolean;

    getId(): number {
        return this.id;
    }
}

function withLogging(
    _target: unknown,
    key: string,
    descriptor: PropertyDescriptor,
): PropertyDescriptor {
    const original = descriptor.value;
    descriptor.value = function (...args: unknown[]) {
        console.log(`Calling ${key}`);
        return original.apply(this, args);
    };
    return descriptor;
}

export class ShopItem extends BaseEntity implements Item, Cacheable {
    public name: string;
    public price: Price;
    public quantity: number;
    public status: ItemStatus;
    public tags?: string[];
    private _cache: Map<string, unknown> = new Map();
    static defaultCurrency = 'USD';

    constructor(id: ItemId, name: string, price: Price) {
        super(id);
        this.name = name;
        this.price = price;
        this.quantity = 0;
        this.status = ItemStatus.Pending;
    }

    validate(): boolean {
        return this.price.amount >= 0 && this.quantity >= 0;
    }

    @withLogging
    updatePrice(newPrice: Price): void {
        this.price = newPrice;
        this._cache.clear();
    }

    async fetchDetails(): Promise<Record<string, unknown>> {
        const result = await utils.fetch(`/items/${this.id}`);
        return result;
    }

    getCacheKey(): string {
        return `item:${this.id}`;
    }

    invalidate(): void {
        this._cache.clear();
    }

    serialize(): Record<string, unknown> {
        return { id: this.id, name: this.name, price: this.price };
    }

    deserialize(data: Record<string, unknown>): void {
        this.name = data['name'] as string;
    }

    static create(name: string, price: Price): ShopItem {
        return new ShopItem(Date.now(), name, price);
    }
}

export class Store<T extends Item = Item> extends EventEmitter {
    private repo: Repository<T>;
    protected cache: Map<ItemId, T> = new Map();

    constructor(repo: Repository<T>) {
        super();
        this.repo = repo;
    }

    async add(item: T): Promise<void> {
        await this.repo.save(item);
        this.cache.set(item.id, item);
        this.emit('added', item);
    }

    async get(id: ItemId): Promise<T | null> {
        if (this.cache.has(id)) {
            return this.cache.get(id) ?? null;
        }
        return this.repo.findById(id);
    }

    protected clearCache(): void {
        this.cache.clear();
    }
}

export function discount(price: Price, pct: number): Price {
    return { ...price, amount: price.amount * (1 - pct / 100) };
}

export function createStore<T extends Item>(repo: Repository<T>): Store<T> {
    return new Store(repo);
}

function formatPrice(price: Price): string {
    return `${price.currency}${price.amount.toFixed(2)}`;
}

File: graph/src/languages/test/fixtures/shop.py
"""Shop module for managing inventory and orders."""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Protocol

logger = logging.getLogger(__name__)

# Module-level constants
STATUS_PENDING = "pending"
STATUS_CONFIRMED = "confirmed"
MAX_ITEMS = 100

# Module-level variable
_registry: dict = {}


class ItemNotFoundError(Exception):
    """Raised when an item cannot be found."""
    pass


class Repository(Protocol):
    """Data-access protocol for item storage."""

    def find_by_id(self, item_id: int) -> Optional["Item"]:
        ...

    def save(self, item: "Item") -> None:
        ...

    def delete(self, item_id: int) -> bool:
        ...


@dataclass
class Item:
    """A product in the shop."""

    name: str
    price: float
    quantity: int = 0
    tags: list = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.price < 0:
            raise ValueError("price must be non-negative")

    @property
    def is_available(self) -> bool:
        """True when the item has stock."""
        return self.quantity > 0

    @classmethod
    def create(cls, name: str, price: float) -> "Item":
        """Factory method."""
        return cls(name=name, price=price)

    @staticmethod
    def validate_price(price: float) -> bool:
        """Validate that a price is non-negative."""
        return price >= 0


class Store:
    """Manages shop inventory."""

    def __init__(self, repo: Repository, name: str = "Default") -> None:
        self.repo = repo
        self.name = name
        self._cache: dict = {}
        self._created_at = datetime.now()

    def add(self, item: Item) -> None:
        """Add an item to the store."""
        if item is None:
            raise ValueError("item cannot be None")
        self.repo.save(item)
        self._cache[id(item)] = item
        logger.info("Added item: %s", item.name)

    def get(self, item_id: int) -> Optional[Item]:
        """Retrieve an item by ID."""
        if item_id in self._cache:
            return self._cache[item_id]
        return self.repo.find_by_id(item_id)

    async def refresh(self) -> None:
        """Asynchronously clear the cache."""
        await asyncio.sleep(0)
        self._cache.clear()

    def _validate(self, item: Item) -> bool:
        """Private validation helper."""
        return item.price >= 0 and item.quantity >= 0


def discount(price: float, pct: float) -> float:
    """Apply a discount percentage to a price."""
    return price * (1 - pct / 100)


async def fetch_prices(ids: list) -> dict:
    """Asynchronously fetch prices for multiple items."""
    await asyncio.sleep(0)
    return {i: 0.0 for i in ids}


def _format_price(price: float) -> str:
    return f"${price:.2f}"
